File: java/mission-to-mars/src/main/java/mars/Rover.java
package mars;

import lombok.Getter;

public class Rover implements ModuleMartien {

	private Direction direction;
	@Getter
	protected Position position;
	@Getter
	private boolean pretPourRecuperation = false;

	public Rover(Direction direction, Position position) {
		this.direction = direction;
		this.position = position;
	}

	public void avancer() {
		switch (direction) {
		case NORD:
			position = new Position(position.getX(), position.getY() + 1, position.getZ());
			break;
		case SUD:
			position = new Position(position.getX(), position.getY() - 1, position.getZ());
			break;
		case OUEST:
			position = new Position(position.getX() - 1, position.getY(), position.getZ());
			break;
		default:
			position = new Position(position.getX() + 1, position.getY(), position.getZ());
		}
	}

	public void reculer() {
		switch (direction) {
		case NORD:
			position = new Position(position.getX(), position.getY() - 1, position.getZ());
			break;
		case SUD:
			position = new Position(position.getX(), position.getY() + 1, position.getZ());
			break;
		case OUEST:
			position = new Position(position.getX() + 1, position.getY(), position.getZ());
			break;
		default:
			position = new Position(position.getX() - 1, position.getY(), position.getZ());
		}
	}

	@Override
	public void monter() {
		throw new UnsupportedOperationException("Un rover ne vole pas");
	}

	@Override
	public void descendre() {
		throw new UnsupportedOperationException("Un rover ne vole pas");
	}

	public void activerRecuperation() {
		pretPourRecuperation = true;
	}

}

File: python/mission-to-mars/src/mars/Rover.py
from src.mars.Direction import Direction
from src.mars.ModuleMartien import ModuleMartien
from src.mars.Position import Position


class Rover(ModuleMartien):

    def __init__(self, direction: Direction, position: Position):
        self._direction = direction
        self.position = position
        self._pret_pour_recuperation: bool = False

    def get_position(self) -> Position:
        return self.position

    def is_pret_pour_recuperation(self) -> bool:
        return self._pret_pour_recuperation

    def avancer(self) -> None:
        if self._direction == Direction.NORD:
            self.position = Position(self.position.x, self.position.y + 1, self.position.z)
        elif self._direction == Direction.SUD:
            self.position = Position(self.position.x, self.position.y - 1, self.position.z)
        elif self._direction == Direction.OUEST:
            self.position = Position(self.position.x - 1, self.position.y, self.position.z)
        else:
            self.position = Position(self.position.x + 1, self.position.y, self.position.z)

    def reculer(self) -> None:
        if self._direction == Direction.NORD:
            self.position = Position(self.position.x, self.position.y - 1, self.position.z)
        elif self._direction == Direction.SUD:
            self.position = Position(self.position.x, self.position.y + 1, self.position.z)
        elif self._direction == Direction.OUEST:
            self.position = Position(self.position.x + 1, self.position.y, self.position.z)
        else:
            self.position = Position(self.position.x - 1, self.position.y, self.position.z)

    def monter(self) -> None:
        raise NotImplementedError("Un rover ne vole pas")

    def descendre(self) -> None:
        raise NotImplementedError("Un rover ne vole pas")

    def activer_recuperation(self) -> None:
        self._pret_pour_recuperation = True
